### ISSUE TITLE
JavaScript heap out of memory error during expo publish

### ISSUE THREAD

**Issue by maxbla** [expo/expo-cli#94]
## Description

I can't publish my app  because `expo publish` gives the following error `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`

## Expected Behavior

Running `expo publish`, I expect some nice messages like `Building iOS bundle` and at the end, something like
```
Uploading JavaScript bundles
Published
Your URL is

<redacted>
```

## Observed Behavior

Node.js ran out of memory. It seems like node's default limit is somewhere around 1.5GB.

```
$ expo publish
[14:16:11] Unable to find an existing Expo CLI instance for this directory, starting a new one...
[14:16:15] Starting Metro Bundler on port 19001.
[14:16:16] Metro Bundler ready.
[14:16:17] Tunnel ready.
[14:16:17] Publishing to channel 'default'...
[14:16:18] Building iOS bundle
[14:19:28] Building Android bundle
[14:19:28] Finished building JavaScript bundle in 190125ms.
[14:23:25] Finished building JavaScript bundle in 236273ms.
[14:23:25] Analyzing assets
[14:28:25] Finished building JavaScript bundle in 300375ms.
[14:39:36] <--- Last few GCs --->
[14:39:36] [7227:0x3570be0]  1400461 ms: Scavenge 1314.5 (1424.6) -> 1313.7 (1424.6) MB, 6.7 / 0.0 ms  (average mu = 0.355, current mu = 0.327) allocation failure
[14:39:36] [7227:0x3570be0]  1400476 ms: Scavenge 1314.6 (1424.6) -> 1313.8 (1424.6) MB, 6.3 / 0.0 ms  (average mu = 0.355, current mu = 0.327) allocation failure
[14:39:36] [7227:0x3570be0]  1400488 ms: Scavenge 1314.7 (1424.6) -> 1313.9 (1425.1) MB, 6.1 / 0.0 ms  (average mu = 0.355, current mu = 0.327) allocation failure
[14:39:36] <--- JS stacktrace --->
[14:39:36] ==== JS stack trace =========================================
[14:39:36] 0: ExitFrame [pc: 0x32254075be1d]
[14:39:36] 1: StubFrame [pc: 0x322540753d47]
[14:39:36] Security context: 0x191c4149e6e1 <JSObject>
[14:39:36] 2: split [0x191c414906c1](this=0x3a0f687af151 <String[1]: :>,0x3e938b29b899 <JSRegExp <String[5]: \r?\n>>)
[14:39:36] 3: print(aka print) [0x1dc22eaaa441] [0x0aa0bf0026f1 <undefined>:~5771] [pc=0x32254591cfea](this=0x0aa0bf0026f1 <undefined>,str=0x3a0f687af151 <String[1]: :>)
[14:39:36] 4: colon [0x2cba29192231] [0x0aa0bf0026f1 <u...
[14:39:36] FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
[14:39:36]  1: 0x8daaa0 node::Abort() [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36]  2: 0x8daaec  [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36] 3: 0xad79ae v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36]  4: 0xad7be4 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36] 5: 0xece9e2  [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36]  6: 0xeceae8 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36]  7: 0xedac12 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36]  8: 0xedb5d4 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36]  9: 0xede241 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36] 10: 0xea7564 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36] 11: 0x114bc12 v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/home/max/.nvm/versions/node/v10.12.0/bin/node]
[14:39:36] 12: 0x32254075be1d
[14:39:42] Error: socket hang up
[14:39:42] RequestError: Error: socket hang up
    at new RequestError (/usr/lib/node_modules/expo-cli/node_modules/request-promise-core/lib/errors.js:14:15)
    at Request.plumbing.callback (/usr/lib/node_modules/expo-cli/node_modules/request-promise-core/lib/plumbing.js:87:29)
    at Request.RP$callback [as _callback] (/usr/lib/node_modules/expo-cli/node_modules/request-promise-core/lib/plumbing.js:46:31)
    at self.callback (/usr/lib/node_modules/expo-cli/node_modules/request/request.js:185:22)
    at Request.emit (events.js:182:13)
    at Request.onRequestError (/usr/lib/node_modules/expo-cli/node_modules/request/request.js:881:8)
    at ClientRequest.emit (events.js:182:13)
    at Socket.socketOnEnd (_http_client.js:425:9)
    at Socket.emit (events.js:187:15)
    at endReadableNT (_stream_readable.js:1094:12)
Building JavaScript bundle [==================================             ] 72%
```

## Environment

```
expo diagnostics
[14:21:59] Generating diagnostics report...
[14:21:59] You can join our slack here: https://slack.expo.io/.

Environment:
  OS: Linux 4.18
  Node: 10.11.0
  Yarn: 1.10.1
  npm: 6.4.1
  Watchman: 4.9.0
  Xcode: N/A
  Android Studio: Not Found

Packages: (wanted => installed)
  expo: ^30.0.0 => 30.0.1
  react: 16.3.1 => 16.3.1
  react-native: https://github.com/expo/react-native/archive/sdk-30.0.0.tar.gz => 0.55.4

Diagnostics report:
  https://exp-xde-diagnostics.s3.amazonaws.com/vaspernaut-381d5d67-08c5-40e4-96fa-a6cdc54a173d.tar.gz
```

Also specify:

- No phone was used

## Reproducible Demo

This doesn't happen with a new project - makes sense - if a new project ran out of memory, I wouldn't be the first to file an issue. I tried adding a few things (assets at first, then a few javascript files) from my project to see if they made it run out of memory, but I couldn't reproduce the problem this way. Due to dependencies, my project can't be split up very easily for testing. I'm not sure how to help people reproduce this bug.

## How have you tried to fix this

Delete some caches. I followed [this](https://forums.expo.io/t/how-to-clear-the-react-native-packager/1352) guide after seeing [this](https://github.com/facebook/react-native/issues/20118)

**Comment by dmdeklerk** [comment]
You can use `export NODE_OPTIONS=--max_old_space_size=4096` since node v8.0 instead.
https://github.com/endel/increase-memory-limit/issues/19

**Comment by DarthBrento** [comment]
> You can use `export NODE_OPTIONS=--max_old_space_size=4096` since node v8.0 instead.
> [endel/increase-memory-limit#19](https://github.com/endel/increase-memory-limit/issues/19)

I get this error simply by running the CLI sometimes after a while. Where do I put this command using PowerShell on Windows 10?

**Comment by mikeRChambers610** [comment]
@dmdeklerk can you please provide more information how we are supposed to "use" the export NODE_OPTIONS=--max_old_space_size=4096? Do we place it somewhere in package.json?

More details would really help. Thanks in advance.

**Comment by maxbla** [comment]
@mikeRChambers610 that is an [environment variable](https://wiki.archlinux.org/index.php/Environment_Variables). It goes in Bash (or your shell of choice) on Linux/macOS. On windows, set environmental variables like [this](https://stackoverflow.com/questions/9249830/how-can-i-set-node-env-production-on-windows).

**Comment by mikeRChambers610** [comment]
This issue seems to be blocking me from uploading my app to the app store. using the max old space is only a work around for development. Still getting this issue a year later...

INFO
19:38
Building JavaScript bundle: 99%
INFO
19:37
transform[stdout]: 
transform[stdout]: <--- Last few GCs --->
transform[stdout]: 
transform[stdout]: [9752:0x104000000]   113041 ms: Scavenge 1224.4 (1431.8) -> 1216.1 (1435.3) MB, 20.1 / 0.0 ms  (average mu = 0.318, current mu = 0.355) allocation failure 
transform[stdout]: [9752:0x104000000]   113131 ms: Scavenge 1227.9 (1435.3) -> 1219.6 (1439.8) MB, 18.8 / 0.0 ms  (average mu = 0.318, current mu = 0.355) allocation failure 
transform[stdout]: [9752:0x104000000]   113239 ms: Scavenge 1231.6 (1439.8) -> 1223.0 (1444.8) MB, 16.2 / 0.0 ms  (average mu = 0.318, current mu = 0.355) allocation failure 
transform[stdout]: 
INFO
19:37
transform[stdout]: 
transform[stdout]: <--- JS stacktrace --->
transform[stdout]: 
transform[stdout]: ==== JS stack trace =========================================
transform[stdout]: 
transform[stdout]:     0: ExitFrame [pc: 0x1c5eeadbe3d]
transform[stdout]: Security context: 0x3eb5fd19e6e9 <JSObject>
transform[stdout]:     1: checkPath [0x3eb57dc0b201] [/Users/mikechambers/exponavs/GroupAid_AWS_CFS_PhoneVerif/NavigationLayout/node_modules/@babel/traverse/lib/path/lib/virtual-types.js:~109] [pc=0x1c5ef0456a4](this=0x3eb57dc0a751 <Object map = 0x3eb5e4db82e1>,path=0x3eb5cb65ef39 <NodePath map = 0x3eb5ede11ea9>)
transform[stdout]:     2: arguments adaptor frame: 2->1
transform[stdout]:     3: visitQueue [0x3eb5b7e...
transform[stdout]: 
INFO
19:37
transform[stderr]: FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
INFO
19:37
transform[stderr]:  1: 0x10003d035 node::Abort() [/usr/local/bin/node]
INFO
19:37
transform[stderr]:  2: 0x10003d23f node::OnFatalError(char const*, char const*) [/usr/local/bin/node]
INFO
19:37
transform[stderr]:  3: 0x1001b8e15 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
INFO
19:37
transform[stderr]:  4: 0x100586d72 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]
INFO
19:37
transform[stderr]:  5: 0x100589845 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [/usr/local/bin/node]
INFO
19:37
transform[stderr]:  6: 0x1005856ef v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/node]
INFO
19:37
transform[stderr]:  7: 0x1005838c4 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
INFO
19:37
transform[stderr]:  8: 0x10059015c v8::internal::Heap::AllocateRawWithLigthRetry(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/usr/local/bin/node]
INFO
19:37
transform[stderr]:  9: 0x1005901df v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [/usr/local/bin/node]
INFO
19:37
transform[stderr]: 10: 0x10055fb24 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [/usr/local/bin/node]
INFO
19:37
transform[stderr]: 11: 0x1007e7e04 v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [/usr/local/bin/node]
INFO
19:37
transform[stderr]: 12: 0x1c5eeadbe3d 
INFO
19:37
transform[stderr]: 13: 0x1c5ef0456a4 
INFO
19:37
transform[stderr]: 14: 0x1c5eea8a5c3 
INFO
19:37
transform[stderr]: 15: 0x1c5eee21821

**Comment by marcosadsj** [comment]
Guys, for some reason typescript module was getting infinity calls on expo build, causing non normal increase memory head use. I just removed it with `yarn remove typescript` and changed all `.tsx` files to `.js` (removing typescript code) and `expo build` worked perfectly! hope helps...

**Comment by EvanBacon** [comment]
We have rolled a potential fix for this into prod for SDK 40, you can try it now with `EXPO_USE_DEV_SERVER=true` which will skip over react-native-community/cli and use Metro directly. If issues persist please feel free to reopen or open a new issue.

**Comment by Rico36** [comment]
where do we enter EXPO_USE_DEV_SERVER=true ?

**Comment by lorstenoplo** [comment]
I found out that one easy way of solving the problem is increasing available space for node and importantly using the root terminal ( like cmd or powershell on windows )

**Comment by anabeatrizzz** [comment]
@lorstenoplo how can I do this?

**Comment by anabeatrizzz** [comment]
@lorstenoplo how can I do this?

**Comment by lorstenoplo** [comment]
@anabeatrizzz For increasing memory allocation for node
 you can do this : 
`$ node --max-old-space-size=2048 index.js`

Check this so question for more info : https://stackoverflow.com/questions/38558989/node-js-heap-out-of-memory

The above command will increase memory allocated to node to 2gb, but check available memory before doing the above. One more option is to set a env var.

**Comment by anabeatrizzz** [comment]
@lorstenoplo I don't have 2GB in my laptop. I ran `expo build` two times and I didn't see this error, I don't know why it's appearing now.

**Comment by lorstenoplo** [comment]
@anabeatrizzz then try with 1gb, btw which terminal are you using

**Comment by anabeatrizzz** [comment]
@lorstenoplo Linux Mint terminal

**Comment by lorstenoplo** [comment]
Is it your root terminal, plus close all other stuff running

**Comment by brentvatne** [comment]
@lsps9150414 - what does your package.json look like?

**Comment by lsps9150414** [comment]
@brentvatne Hey, no I don't have dependency on that:

```
{
  "main": "node_modules/expo/AppEntry.js",
  "private": true,
  "scripts": {
    "switch-env:development": "cp google-services.development.json google-services.json && cp GoogleService-Info.development.plist GoogleService-Info.plist && cp .env.development .env",
    "switch-env:staging": "cp google-services.staging.json google-services.json && cp GoogleService-Info.staging.plist GoogleService-Info.plist && cp .env.staging .env",
    "switch-env:production": "cp google-services.production.json google-services.json && cp GoogleService-Info.production.plist GoogleService-Info.plist && cp .env.production .env",
    "android": "expo start --android",
    "ios": "expo start --ios",
    "web": "expo start --web",
    "start": "REACT_DEBUGGER='rndebugger-open --open --expo' expo start",
    "start:local": "ENV=local yarn start",
    "start:development": "yarn switch-env:development && yarn start",
    "start:staging": "yarn switch-env:staging && yarn start",
    "start:production": "yarn switch-env:production && yarn start",
    "lint": "eslint",
    "tsc": "tsc --noEmit",
    "test": "jest --passWithNoTests",
    "version-bump": "scripts/bumpAppVersion.sh"
  },
  "dependencies": {
    "@bit/lsps9150414.utils.array-to-object": "0.0.5",
    "@bit/lsps9150414.utils.async-action": "0.0.1",
    "@bit/lsps9150414.utils.composite-key-generator": "0.0.1",
    "@bit/lsps9150414.utils.create-action-type-guard": "0.0.1",
    "@bit/lsps9150414.utils.create-saga-observer": "0.0.1",
    "@bit/lsps9150414.utils.create-saga-task": "0.0.1",
    "@bit/lsps9150414.utils.create-saga-ticker": "0.0.1",
    "@bit/lsps9150414.utils.create-saga-watcher": "0.0.1",
    "@bit/lsps9150414.utils.create-typed-reducer": "0.0.1",
    "@bit/lsps9150414.utils.data-validators": "0.0.1",
    "@bit/lsps9150414.utils.firebase-firestore-helper": "0.0.1",
    "@bit/lsps9150414.utils.firebase-storage-helper": "0.0.1",
    "@bit/lsps9150414.utils.firebase.get-moment-from-firestore-timestamp": "0.0.1",
    "@bit/lsps9150414.utils.firebase.schema": "0.0.1",
    "@bit/quaza.utils.firestore-paths": "0.0.1",
    "@bit/quaza.utils.get-new-inventory-amount": "^0.0.3",
    "@bit/quaza.utils.is-profile-valid": "0.0.2",
    "@expo/metro-config": "^0.1.59",
    "@expo/react-native-action-sheet": "^3.8.0",
    "@ptomasroos/react-native-multi-slider": "^2.2.2",
    "@react-native-community/async-storage": "~1.12.0",
    "@react-native-community/masked-view": "0.1.10",
    "@react-navigation/bottom-tabs": "^5.11.2",
    "@react-navigation/native": "^5.8.10",
    "@react-navigation/stack": "^5.12.8",
    "@reduxjs/toolkit": "^1.0.4",
    "@types/bluebird": "^3.5.33",
    "@types/validator": "^13.1.1",
    "algoliasearch": "^4.5.1",
    "axios": "^0.19.0",
    "bluebird": "^3.7.2",
    "buffer": "^5.6.0",
    "dotenv": "^8.2.0",
    "expo": "^40.0.0",
    "expo-analytics-segment": "~9.1.0",
    "expo-apple-authentication": "~2.2.2",
    "expo-application": "~2.4.1",
    "expo-constants": "~9.3.3",
    "expo-crypto": "~8.4.0",
    "expo-facebook": "~9.1.0",
    "expo-firebase-analytics": "~2.6.0",
    "expo-firebase-core": "~1.3.0",
    "expo-image-picker": "~9.2.0",
    "expo-intent-launcher": "~8.4.0",
    "expo-linear-gradient": "~8.4.0",
    "expo-linking": "~2.0.1",
    "expo-location": "~10.0.0",
    "expo-notifications": "~0.8.2",
    "expo-permissions": "~10.0.0",
    "expo-sharing": "~8.5.0",
    "expo-splash-screen": "~0.8.1",
    "expo-status-bar": "~1.0.3",
    "expo-updates": "~0.4.1",
    "expo-web-browser": "~8.6.0",
    "firebase": "7.9.0",
    "firestore-size": "^2.0.5",
    "geolib": "^3.1.0",
    "hex-rgba": "^1.0.2",
    "lodash": "^4.17.15",
    "lottie-react-native": "~2.6.1",
    "moment": "^2.24.0",
    "numeral": "^2.0.6",
    "react": "16.13.1",
    "react-dom": "16.13.1",
    "react-native": "0.63.4",
    "react-native-animatable": "^1.3.3",
    "react-native-drag-sort": "^2.3.7",
    "react-native-elements": "^1.2.7",
    "react-native-gesture-handler": "~1.8.0",
    "react-native-gifted-chat": "^0.16.3",
    "react-native-keyboard-aware-scroll-view": "^0.9.2",
    "react-native-modal": "^11.5.6",
    "react-native-progress-circle": "^2.1.0",
    "react-native-reanimated": "~1.13.0",
    "react-native-safe-area-context": "3.1.9",
    "react-native-screens": "~2.15.2",
    "react-native-snap-carousel": "^4.0.0-beta.5",
    "react-native-svg": "12.1.0",
    "react-native-swiper": "^1.6.0",
    "react-native-unimodules": "~0.12.0",
    "react-native-view-shot": "3.1.2",
    "react-native-web": "~0.13.12",
    "react-redux": "^7.1.3",
    "react-scripts": "^4.0.3",
    "redux": "^4.0.4",
    "redux-actions": "^2.6.5",
    "redux-persist": "^6.0.0",
    "redux-saga": "^1.1.3",
    "reselect": "^4.0.0",
    "sentry-expo": "^3.0.2",
    "tinycolor2": "^1.4.1",
    "tinygradient": "^1.1.1",
    "uuid": "^3.3.3",
    "validator": "^13.0.0"
  },
  "devDependencies": {
    "@babel/core": "~7.9.0",
    "@expo/config": "^3.3.9",
    "@types/hex-rgba": "^1.0.0",
    "@types/lodash": "^4.14.150",
    "@types/react": "~16.9.35",
    "@types/react-dom": "~16.9.8",
    "@types/react-native": "~0.63.2",
    "@types/react-redux": "^7.1.7",
    "@types/redux-actions": "^2.6.1",
    "@typescript-eslint/eslint-plugin": "^4.20.0",
    "@typescript-eslint/parser": "^4.20.0",
    "babel-plugin-module-resolver": "^4.0.0",
    "eslint": "^7.2.0",
    "eslint-config-airbnb": "18.2.1",
    "eslint-config-prettier": "^8.1.0",
    "eslint-import-resolver-babel-module": "^5.1.2",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-jsx-a11y": "^6.4.1",
    "eslint-plugin-prettier": "^3.3.1",
    "eslint-plugin-react": "^7.21.5",
    "eslint-plugin-react-hooks": "^1.7.0",
    "eslint-plugin-react-native": "^3.10.0",
    "expo-cli": "^4.3.2",
    "git-semver-tags": "^4.1.1",
    "husky": "^3.0.9",
    "import-sort": "^6.0.0",
    "import-sort-style-absolute": "^1.0.1",
    "jest-expo": "^40.0.0",
    "jetifier": "^1.6.4",
    "lint-staged": "^10.4.0",
    "prettier": "^2.2.1",
    "react-native-debugger-open": "^0.3.25",
    "react-native-svg-transformer": "^0.14.3",
    "standard-version": "^9.0.0",
    "standard-version-expo": "^1.0.3",
    "typescript": "~4.0.0"
  },
  "jest": {
    "preset": "react-native"
  },
  "lint-staged": {
    "*.{js,jsx,ts,tsx}": [
      "eslint --fix",
      "bash -c tsc --noEmit"
    ],
    "*.{json,css,md}": [
      "prettier --write"
    ]
  },
  "husky": {
    "hooks": {
      "pre-commit": "lint-staged"
    }
  },
  "importSort": {
    ".js, .jsx, .ts, .tsx, .es6, .es": {
      "parser": "babylon",
      "style": "absolute"
    }
  },
  "bit": {
    "env": {
      "compiler": {
        "bit.envs/compilers/typescript@3.1.52": {
          "rawConfig": {
            "tsconfig": {
              "compilerOptions": {
                "target": "es2015",
                "module": "CommonJS"
              }
            }
          }
        }
      }
    },
    "componentsDefaultDirectory": "components/{name}",
    "packageManager": "yarn"
  }
}
```

**Comment by lsps9150414** [comment]
@brentvatne 

Using the following command worked:
```
EXPO_USE_DEV_SERVER=true NODE_OPTIONS=--max_old_space_size=2048 npx expo build:ios --non-interactive --skip-workflow-check --skip-credentials-check -t archive --release-channel "MY_CHANNEL_NAME" --no-wait
```

But IMHO ideally we shouldn't need to add `NODE_OPTIONS=--max_old_space_size=2048`?
Also, I don't know if `EXPO_USE_DEV_SERVER=true` did anything.

**Comment by brentvatne** [comment]
@lsps9150414 - you should open an issue on https://github.com/facebook/metro about it and provide them with a reproducible example :)

**Comment by jmarks-joshua** [comment]
I have just been getting this issue, but my app is really quite small so I didn't really believe it was anything to do with being too large, I could also see from the logs that it was getting to 4gb. I've finally just found the issue, which was that somehow my IDE had added an `import { clearConfigCache } from 'prettier'` line not sure why that caused an infinite loop but removing it seems to have done the job.

**Comment by isocra** [comment]
I now get the JavaScript heap out of memory error all the time and am unable to publish our app at all!!! I'm trying it locally and with our Jenkins CI server. 

Based on what I've read above, The command I'm using is:
```
EXPO_USE_DEV_SERVER=true NODE_OPTIONS=--max_old_space_size=4096 yarn expo publish --release-channel develop
```
The specific error I get is `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`. I've tried adding `--clear` but it doesn't make any difference.

@brentvatne asked earlier of someone was including the `aws-sdk` and we are, we need it. 

Any other ideas for work-arounds and fixes? I also find that `yarn expo start` is very slow, like more than 10 mins on my Mac M1. What could be going wrong?

Here are my local diagnostics:
```
  Expo CLI 4.9.1 environment info:
    System:
      OS: macOS 11.5.2
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 14.16.0 - /var/folders/x_/4sk4s5cn50g7jx0ymb9f7lmr0000gp/T/yarn--1632407554719-0.1811081058863715/node
      Yarn: 1.22.10 - /var/folders/x_/4sk4s5cn50g7jx0ymb9f7lmr0000gp/T/yarn--1632407554719-0.1811081058863715/yarn
      npm: 6.14.11 - ~/.nvm/versions/node/v14.16.0/bin/npm
      Watchman: 4.9.0 - /opt/homebrew/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
      Android SDK:
        API Levels: 27, 28
        Build Tools: 28.0.2, 28.0.3
        System Images: android-28 | Google Play Intel x86 Atom, android-Q | Google APIs Intel x86 Atom
    IDEs:
      Android Studio: 3.4 AI-183.5429.30.34.5452501
      Xcode: 12.5.1/12E507 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^42.0.0 => 42.0.1 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-42.0.0.tar.gz => 0.63.2 
      react-native-web: ^0.17.1 => 0.17.1 
    npmGlobalPackages:
      expo-cli: 4.9.1
    Expo Workflow: managed
```

**Comment by brentvatne** [comment]
@isocra - you can try `EXPO_PROFILE=1 expo start` to get an idea of what is taking a long time. are you possibly inside of a massive monorepo? metro (react native's built-in bundler) indexes the entire repository before starting, so that could be part of the issue. as far as memory usage caused by aws-sdk goes, there isn't much we can do here as the responsibility here lives in both metro and aws-sdk. you can try to bump the memory limit even higher

**Comment by isocra** [comment]
Thanks for the quick reply @brentvatne! I tried `EXPO_PROFILE=1 yarn expo start` as you suggested and got some timing info, but it stopped outputting after `[profile] TerminalUI.startAsync: 88.688ms`. All that bit was fast, then there's a warning:
```
jest-haste-map: Watchman crawl failed. Retrying once with node crawler.
  Usually this happens when watchman isn't running. Create an empty `.watchmanconfig` file in your project's root folder or initialize a git or hg repository in your project.
  Error: Watchman error: too many pending cache jobs. Make sure watchman is running for this project. See https://facebook.github.io/watchman/docs/troubleshooting.html.
```
This is quite annoying because I _do_ have a `.watchmanconfig` file in the current directory, which is also the root of my project. It contains just `{}`. But doesn't prevent the error message.

Then when I try and connect a device (iOS or Android), the device says downloading... but nothing happens for some minutes. Then eventually I get the command-line progress bar and things start to work. The compilation/build step with the progress bar is quite fast (< 2mins I guess), but it's the weird pause, sometimes for many minutes, before it starts that I can't explain. Quite often the device gives up and says it can't connect to the server so I have to tell it to try again.

**Comment by MoSattler** [comment]
My project was running fine for a long time - and all of the sudden it stopped working. I don't have ANY assets yet that aren't there by default.

**Comment by iamshadmirza** [comment]
Getting this issue with GitHub action, tried everything listed in thread above. More info in screenshot below.

> Note: `eas build` is working fine, only `expo publish` has this issue
<img width="1206" alt="Screenshot 2022-02-01 at 5 13 25 PM" src="https://user-images.githubusercontent.com/22999030/151962592-1f8ef8bf-3c9a-4cfe-afae-e1568e9f901b.png">

**Comment by foxfl** [comment]
> Getting this issue with GitHub action, tried everything listed in thread above. More info in screenshot below.
> 
> > Note: `eas build` is working fine, only `expo publish` has this issue
> 
> <img alt="Screenshot 2022-02-01 at 5 13 25 PM" width="1206" src="https://user-images.githubusercontent.com/22999030/151962592-1f8ef8bf-3c9a-4cfe-afae-e1568e9f901b.png">

+1

**Comment by bugcloud** [comment]
> Getting this issue with GitHub action,

I have same issue.  
My GitHub action pass this issue by changing `runs-on` config from `ubuntu-latest` to `macos-latest` .

**Comment by bellalMohamed** [comment]
> Getting this issue with GitHub action, tried everything listed in thread above. More info in screenshot below.
> 
> > Note: `eas build` is working fine, only `expo publish` has this issue
> 
> <img alt="Screenshot 2022-02-01 at 5 13 25 PM" width="1206" src="https://user-images.githubusercontent.com/22999030/151962592-1f8ef8bf-3c9a-4cfe-afae-e1568e9f901b.png">

@foxfl did you manage to fix the issue?
